{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "typical-freight",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "# COBRApy\n",
    "\n",
    "COBRApy is a package for constraint-based modeling of biological networks written in Python.\n",
    "\n",
    "This tool allows loading and inspecting Genome-Scale Metabolic (GEM) models written in the Systems Biology Markup Language (SBML) format.\n",
    "\n",
    "Using COBRApy, one can analyse the following model contents:\n",
    "1. Reactions\n",
    "2. Metabolites\n",
    "3. Genes\n",
    "4. Exchange reactions (Environmental Conditions)\n",
    "\n",
    "COBRApy allows manipulating the contents of a GEM model. For instance, one can edit reactions' flux bounds, knock out a metabolic gene, or change the environmental conditions.\n",
    "\n",
    "Phenotype prediction can be simulated with several flux analysis methods implemented in COBRApy. These include Flux Balance Analysis (FBA), Parsimonious FBA, or Flux Variability Analysis (FVA).\n",
    "\n",
    "The simulation of gene and reaction deletions for a given GEM model is a simple and straightforward process. One can simulate single or double knock outs using one of the flux analysis methods mentioned above."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "boring-compensation",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "## Instalation\n",
    "\n",
    "\n",
    "### Requirements\n",
    "The following requirements need to be installed to use COBRApy:\n",
    "- Python 3.6 or higher\n",
    "- pip must be installed\n",
    "- GLPK solver is used by default but CPLEX is prefered\n",
    "\n",
    "\n",
    "### How to install COBRApy?\n",
    "```\n",
    "pip install cobra\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "political-twenty",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "## Working with a GEM model\n",
    "\n",
    "COBRApy can be used to read a GEM model in SBML format file.\n",
    "For this practical session, we will be using two models:\n",
    "- _E. coli_ core model which contains the central carbon metabolism of _Escherichia coli_ -> file: ..data/e_coli_core.xml\n",
    "\n",
    "The reactions, metabolites, and genes encoded into a SBML format file can be parsed by COBRApy. These contents are loaded into Python objects simple to use, namely `cobra.Reaction`, `cobra.Metabolite`, and `cobra.Gene`\n",
    "\n",
    "The model itself will be available as an `cobra.Model` object containing all these attributes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "minimal-productivity",
   "metadata": {
    "scrolled": true,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>e_coli_core</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x024bed7a8888</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>72</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>95</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*BIOMASS_Ecoli_core_w_GAM - 1.0*BIOMASS_Ecoli_core_w_GAM_reverse_712e5</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>extracellular space, cytosol</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model e_coli_core at 0x24bed7a8888>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# importing cobra\n",
    "import cobra\n",
    "\n",
    "# Loading a model\n",
    "model_path = '../data/e_coli_core.xml'\n",
    "model = cobra.io.read_sbml_model(model_path)\n",
    "\n",
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "271ea401",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<Reaction PFK at 0x24bee86d748>,\n",
       " <Reaction PFL at 0x24bee86d708>,\n",
       " <Reaction PGI at 0x24bee874a88>,\n",
       " <Reaction PGK at 0x24bee87aa08>,\n",
       " <Reaction PGL at 0x24bee87afc8>]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#retriving first five reactions of the model\n",
    "model.reactions[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "db222e49",
   "metadata": {
    "scrolled": true,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>PFK</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Phosphofructokinase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x24bee86d748</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>atp_c + f6p_c --> adp_c + fdp_c + h_c</p>\n",
       "                    <p style='text-align:right'>ATP C10H12N5O13P3 + D-Fructose 6-phosphate --> ADP C10H12N5O10P2 + D-Fructose 1,6-bisphosphate + H+</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>b3916 or b1723</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction PFK at 0x24bee86d748>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#inspecting the first reaction of the model\n",
    "model.reactions[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3204cb7c",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>ACALD</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Acetaldehyde dehydrogenase (acetylating)</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x24bee87af08</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>acald_c + coa_c + nad_c <=> accoa_c + h_c + nadh_c</p>\n",
       "                    <p style='text-align:right'>Acetaldehyde + Coenzyme A + Nicotinamide adenine dinucleotide <=> Acetyl-CoA + H+ + Nicotinamide adenine dinucleotide - reduced</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>b0351 or b1241</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-1000.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction ACALD at 0x24bee87af08>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#inspecting a reaction by its ID\n",
    "model.reactions.get_by_id(\"ACALD\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e0b93231",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<Metabolite glc__D_e at 0x24bed7dffc8>,\n",
       " <Metabolite gln__L_c at 0x24bed7dff88>,\n",
       " <Metabolite gln__L_e at 0x24bed7ee788>,\n",
       " <Metabolite glu__L_c at 0x24bee7c22c8>,\n",
       " <Metabolite glu__L_e at 0x24bee7c2508>]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#retriving first five metabolites of the model\n",
    "model.metabolites[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d31e05e5",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Metabolite identifier</strong></td><td>glc__D_e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>D-Glucose</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x24bed7dffc8</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Formula</strong></td><td>C6H12O6</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartment</strong></td><td>e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>In 2 reaction(s)</strong></td><td>\n",
       "                    EX_glc__D_e, GLCpts\n",
       "                    </td>\n",
       "            </tr>\n",
       "        </table>"
      ],
      "text/plain": [
       "<Metabolite glc__D_e at 0x24bed7dffc8>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#inspecting the first metabolite of the model\n",
    "model.metabolites[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1992e8f8",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Metabolite identifier</strong></td><td>13dpg_c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>3-Phospho-D-glyceroyl phosphate</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x24bee7d2408</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Formula</strong></td><td>C3H4O10P2</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartment</strong></td><td>c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>In 2 reaction(s)</strong></td><td>\n",
       "                    PGK, GAPD\n",
       "                    </td>\n",
       "            </tr>\n",
       "        </table>"
      ],
      "text/plain": [
       "<Metabolite 13dpg_c at 0x24bee7d2408>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#inspecting a metabolite by its ID\n",
    "model.metabolites.get_by_id(\"13dpg_c\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "83287794",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<Gene b1241 at 0x24bee8049c8>,\n",
       " <Gene b0351 at 0x24bee8047c8>,\n",
       " <Gene s0001 at 0x24bed7bbf48>,\n",
       " <Gene b1849 at 0x24bee814608>,\n",
       " <Gene b3115 at 0x24bee814a88>]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#retriving first five genes of the model\n",
    "model.genes[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b4683d3a",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Gene identifier</strong></td><td>b1241</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>adhE</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x24bee8049c8</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Functional</strong></td><td>True</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>In 2 reaction(s)</strong></td><td>\n",
       "                    ACALD, ALCD2x\n",
       "                    </td>\n",
       "            </tr>\n",
       "        </table>"
      ],
      "text/plain": [
       "<Gene b1241 at 0x24bee8049c8>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#inspecting the first gene of the model\n",
    "model.genes[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7cae917a",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Gene identifier</strong></td><td>b1241</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>adhE</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x24bee8049c8</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Functional</strong></td><td>True</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>In 2 reaction(s)</strong></td><td>\n",
       "                    ACALD, ALCD2x\n",
       "                    </td>\n",
       "            </tr>\n",
       "        </table>"
      ],
      "text/plain": [
       "<Gene b1241 at 0x24bee8049c8>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#inspecting a gene by its ID\n",
    "model.genes.get_by_id('b1241')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0c64159",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "### Reactions\n",
    "\n",
    "In COBRApy reactions are objects that can be inspected. These objects can contain usefull information for all reactions in the model, such as:\n",
    "- name;\n",
    "- metabolites;\n",
    "- stoichiometry;\n",
    "- genes;\n",
    "- formula;\n",
    "- reversibility;\n",
    "- flux bounds;\n",
    "- gpr;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9c5ac10a",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Acetaldehyde dehydrogenase (acetylating) \n",
      "\n",
      "acald_c + coa_c + nad_c <=> accoa_c + h_c + nadh_c \n",
      "\n",
      "acald_c -> -1.0\n",
      "coa_c -> -1.0\n",
      "nad_c -> -1.0\n",
      "accoa_c -> 1.0\n",
      "h_c -> 1.0\n",
      "nadh_c -> 1.0\n"
     ]
    }
   ],
   "source": [
    "#inspecting the reaction name, formula, metabolites, and stoichiometry.\n",
    "reaction = model.reactions.get_by_id('ACALD')\n",
    "\n",
    "print(reaction.name, '\\n')\n",
    "print(reaction.reaction, '\\n')\n",
    "for metabolite, coefficient in reaction.metabolites.items():\n",
    "    print(metabolite, '->', coefficient)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "24b8e81a",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-1000.0 < ACALD < 1000.0 \n",
      "\n",
      "True \n",
      "\n",
      "(-1000.0, 1000.0)\n"
     ]
    }
   ],
   "source": [
    "#inspecting reversibility and flux bounds.\n",
    "print(reaction.lower_bound, \"< ACALD <\", reaction.upper_bound, '\\n')\n",
    "print(reaction.reversibility, '\\n')\n",
    "print(reaction.bounds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7f189664",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-1000.0 < ACALD < 1000.0 \n",
      "\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "#change reaction bounds\n",
    "reaction.bounds = (0, 1000)\n",
    "print(reaction.lower_bound, \"< ACALD <\", reaction.upper_bound, '\\n')\n",
    "print(reaction.reversibility)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "approved-respect",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "#### GPR\n",
    "A Gene-Protein-Reaction (GPR) rule is a boolean representation of the gene requirements for a given reaction to be active.\n",
    "In COBRApy, a string representation of the GPR is stored as the `gene_reaction_rule` in the `Reaction` object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d0941fd8",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'b0351 or b1241'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#inspect gene reaction rule\n",
    "reaction.gene_reaction_rule"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "transsexual-instruction",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "In the `ACALD` reaction, either gene *b0351* or *b1241* are required for this reaction to be active."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66289c1c",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "### Metabolites\n",
    "\n",
    "In COBRApy metabolites are objects that can be inspected. These objects can contain usefull information for all metabolites in the model, such as:\n",
    "- name;\n",
    "- chemical formula;\n",
    "- compartment;\n",
    "- reactions;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "4a10d4d8",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3-Phospho-D-glyceroyl phosphate \n",
      "\n",
      "C3H4O10P2 \n",
      "\n",
      "Metabolite Compartment: c \n",
      "\n",
      "PGK : acald_c + coa_c + nad_c <=> accoa_c + h_c + nadh_c\n",
      "GAPD : acald_c + coa_c + nad_c <=> accoa_c + h_c + nadh_c\n"
     ]
    }
   ],
   "source": [
    "#inspecting metabolite name, chemical formula, compartment and reactions.\n",
    "metabolite = model.metabolites.get_by_id('13dpg_c')\n",
    "\n",
    "print(metabolite.name, '\\n')\n",
    "print(metabolite.formula, '\\n')\n",
    "print('Metabolite Compartment:',metabolite.compartment, '\\n')\n",
    "for _reaction in metabolite.reactions:\n",
    "    print(_reaction.id, ':', reaction.reaction)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b1f9f95",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "### Genes\n",
    "\n",
    "In COBRApy metabolites are objects that can be inspected. These objects can contain usefull information for all metabolites in the model, such as:\n",
    "- name;\n",
    "- reactions;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d276d220",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mhpF \n",
      "\n",
      "ACALD : b0351 or b1241\n"
     ]
    }
   ],
   "source": [
    "#inspect gene name and reactions.\n",
    "gene = model.genes.get_by_id('b0351')\n",
    "\n",
    "print(gene.name, '\\n')\n",
    "for reaction_ in gene.reactions:\n",
    "    print(reaction_.id, ':', reaction.gene_reaction_rule)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2da1dce1",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "### Exchanges"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b061d983",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "In COBRApy exchanges are reaction objects that can be inspected in the model. These special reactions define the environmental condictions (eg. medium) of the model. According to the lower and upper bound of each exchange reactions, one can visualize the model's uptake and secretion metabolites."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "98894b23",
   "metadata": {
    "scrolled": true,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Acetate exchange -> (0.0, 1000.0)\n",
      "Acetaldehyde exchange -> (0.0, 1000.0)\n",
      "2-Oxoglutarate exchange -> (0.0, 1000.0)\n",
      "CO2 exchange -> (-1000.0, 1000.0)\n",
      "Ethanol exchange -> (0.0, 1000.0)\n",
      "Formate exchange -> (0.0, 1000.0)\n",
      "D-Fructose exchange -> (0.0, 1000.0)\n",
      "Fumarate exchange -> (0.0, 1000.0)\n",
      "D-Glucose exchange -> (-10.0, 1000.0)\n",
      "L-Glutamine exchange -> (0.0, 1000.0)\n",
      "L-Glutamate exchange -> (0.0, 1000.0)\n",
      "H+ exchange -> (-1000.0, 1000.0)\n",
      "H2O exchange -> (-1000.0, 1000.0)\n",
      "D-lactate exchange -> (0.0, 1000.0)\n",
      "L-Malate exchange -> (0.0, 1000.0)\n",
      "Ammonia exchange -> (-1000.0, 1000.0)\n",
      "O2 exchange -> (-1000.0, 1000.0)\n",
      "Phosphate exchange -> (-1000.0, 1000.0)\n",
      "Pyruvate exchange -> (0.0, 1000.0)\n",
      "Succinate exchange -> (0.0, 1000.0)\n"
     ]
    }
   ],
   "source": [
    "#inspecting the reaction name, formula, metabolites, and stoichiometry.\n",
    "for exchange in model.exchanges:\n",
    "    print(exchange.name, '->', exchange.bounds)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "female-nursery",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "### Linear problem\n",
    "The number of degrees of freedom can be calculated by subtracting the number of variables by the number of linear mass equations. That is, one can determine the degree of under-determination of this model by subtracting the number reactions by the number of metabolites."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "purple-threat",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "23"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# calculing the linear problem degrees of freedom\n",
    "degrees = len(model.reactions) - len(model.metabolites)\n",
    "degrees"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7de6bd17",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "### Phenotype Prediction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5fea53b7",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "COBRApy includes three different algortithms for phenotype prediction. These include Flux Balance Analysis (FBA), Parsimonious Flux Balance Analysis (pFBA), and Flux Variability Analysis (FVA).\n",
    "\n",
    "To perform a simulation using one of these methods, you should first define an objective function. This can be a reaction or an exchange, which will be maximized or minimized. By default, the biomass reaction is set as the model objective function as this mimics the biological behavior of most organisms."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f026a013",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "model.objective = 'BIOMASS_Ecoli_core_w_GAM'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3462f94e",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "#### Flux Balance Analysis (FBA)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75a8c7f0",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "FBA simulations can be performed using `model.optimize()`. This returns a solution object, which includes the result of the simulation. This includes:\n",
    "- objective_value;\n",
    "- status;\n",
    "- fluxes;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "a612d911",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong><em>Optimal</em> solution with objective value 0.874</strong><br><div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fluxes</th>\n",
       "      <th>reduced_costs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PFK</th>\n",
       "      <td>7.477382</td>\n",
       "      <td>8.673617e-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PFL</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.527746e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGI</th>\n",
       "      <td>4.860861</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGK</th>\n",
       "      <td>-16.023526</td>\n",
       "      <td>3.469447e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGL</th>\n",
       "      <td>4.959985</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NADH16</th>\n",
       "      <td>38.534610</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NADTRHD</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-2.546243e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NH4t</th>\n",
       "      <td>4.765319</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O2t</th>\n",
       "      <td>21.799493</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PDH</th>\n",
       "      <td>9.282533</td>\n",
       "      <td>1.734723e-18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>95 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<Solution 0.874 at 0x24bf314d688>"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#performing a FBA simulation\n",
    "fba_solution = model.optimize()\n",
    "fba_solution"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "326574a5",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Models solved using the FBA method can be further analysed using the `model.summary()` method.\n",
    "\n",
    "\n",
    "This method returns the flux value for the reaction defined as objective function. It also returns the fluxes for the exchange reactions. In this table, one can see uptake fluxes, which corresponds to the rates of metabolite cosumption in the model. On the other hand, the secretion fluxes corresponds to the rates of metabolite production in the model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "89a661f7",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 BIOMASS_Ecoli_core_w_GAM = 0.8739215069684302</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>glc__D_e</td>\n",
       "      <td>EX_glc__D_e</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>100.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e</td>\n",
       "      <td>4.765</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e</td>\n",
       "      <td>21.8</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e</td>\n",
       "      <td>3.215</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e</td>\n",
       "      <td>-22.81</td>\n",
       "      <td>1</td>\n",
       "      <td>100.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e</td>\n",
       "      <td>-29.18</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e</td>\n",
       "      <td>-17.53</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x24beeaced48>"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12014348",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "#### Parsimonious Flux Balance Analysis (pFBA)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abf11c94",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "pFBA simulations gives the optimal growth rate, while minimizing the total sum of fluxes.\n",
    "pFBA can be used from the flux analysis package `cobra.flux_analysis.pfba(model)`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "1b712fd0",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong><em>Optimal</em> solution with objective value 518.422</strong><br><div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fluxes</th>\n",
       "      <th>reduced_costs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PFK</th>\n",
       "      <td>7.477382</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PFL</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.733333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGI</th>\n",
       "      <td>4.860861</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGK</th>\n",
       "      <td>-16.023526</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGL</th>\n",
       "      <td>4.959985</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NADH16</th>\n",
       "      <td>38.534610</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NADTRHD</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.422222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NH4t</th>\n",
       "      <td>4.765319</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O2t</th>\n",
       "      <td>21.799493</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PDH</th>\n",
       "      <td>9.282533</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>95 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<Solution 518.422 at 0x24beead0188>"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#performing pfba simulation\n",
    "pfba_solution = cobra.flux_analysis.pfba(model)\n",
    "pfba_solution"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "426310dc",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "The optimal solution of the pFBA is considerably different from the FBA result. This happens because the objective value for the pFBA is defined as the sum of all flux values (`sum(abs(pfba_solution.fluxes.values))`). On the other hand the FBA result corresponds to the flux value of the reaction that is being optimized (`fba_solution.fluxes[\"Biomass_Ecoli_core\"]`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e416cbc9",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "518.4220855176067"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#calculating the objective value of a pFBA solution\n",
    "sum(abs(pfba_solution.fluxes.values))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15c67909",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "#### Flux Variability Analysis (FVA)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6cc5f62",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "FBA does not obtain an unique flux distribution for a given objective function, but rathar a space with multiple optimum soltuion. FVA is a simulation method that finds the possible flux range for each reaction while optimizing the objective function. FVA can be used from the flux analysis package `cobra.flux_analysis.flux_variability_analysis(model)`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b2b4c04a",
   "metadata": {
    "scrolled": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>minimum</th>\n",
       "      <th>maximum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PFK</th>\n",
       "      <td>7.477382</td>\n",
       "      <td>7.477382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PFL</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGI</th>\n",
       "      <td>4.860861</td>\n",
       "      <td>4.860861</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGK</th>\n",
       "      <td>-16.023526</td>\n",
       "      <td>-16.023526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGL</th>\n",
       "      <td>4.959985</td>\n",
       "      <td>4.959985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NADH16</th>\n",
       "      <td>38.534610</td>\n",
       "      <td>38.534610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NADTRHD</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NH4t</th>\n",
       "      <td>4.765319</td>\n",
       "      <td>4.765319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O2t</th>\n",
       "      <td>21.799493</td>\n",
       "      <td>21.799493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PDH</th>\n",
       "      <td>9.282533</td>\n",
       "      <td>9.282533</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>95 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           minimum    maximum\n",
       "PFK       7.477382   7.477382\n",
       "PFL       0.000000   0.000000\n",
       "PGI       4.860861   4.860861\n",
       "PGK     -16.023526 -16.023526\n",
       "PGL       4.959985   4.959985\n",
       "...            ...        ...\n",
       "NADH16   38.534610  38.534610\n",
       "NADTRHD   0.000000   0.000000\n",
       "NH4t      4.765319   4.765319\n",
       "O2t      21.799493  21.799493\n",
       "PDH       9.282533   9.282533\n",
       "\n",
       "[95 rows x 2 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#performing fva simulation\n",
    "fva_solution = cobra.flux_analysis.flux_variability_analysis(model)\n",
    "fva_solution"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8ad48a1",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "### Simulating Deletions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1aed98f",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "As previously mentioned, COBRApy can be used to simulate gene or reaction knock outs or deletions."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "227c1a25",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "#### Single Reaction and Gene Knock outs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c0a6e67",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "The function `knock_out()` can be used to access what happens when a specific reaction is knocked out and not allowed to have any flux. Considering the previously analysed PFK reaction:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "fbb59894",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PFK knocked out:  <Solution 0.704 at 0x24beec24148>\n"
     ]
    }
   ],
   "source": [
    "#knock out the ACALD reaction\n",
    "with model:\n",
    "    model.reactions.PFK.knock_out()\n",
    "    print('PFK knocked out: ', model.optimize())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "b2ede2bc",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "b1723 knocked out:  <Solution 0.874 at 0x24beec27248>\n",
      "b3916 knocked out:  <Solution 0.704 at 0x24beec12048>\n"
     ]
    }
   ],
   "source": [
    "#knock out the b1723 and b3916 genes, which are associated with the pfk reaction\n",
    "with model:\n",
    "    model.genes.b1723.knock_out()\n",
    "    print('b1723 knocked out: ', model.optimize())\n",
    "    model.genes.b3916.knock_out()\n",
    "    print('b3916 knocked out: ', model.optimize())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92f18d76",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Moreover, COBRApy incorporates to simulation methods that are used predict the flux distribution after a gene knock out. These are the Minimization of Metabolic Adjustment (MOMA), which can be called using `cobra.flux_analysis.moma()`, and Regulatory On/Off Minimization (ROOM), using `cobra.flux_analysis.room()`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "6101d834",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MOMA Result:  <Solution 288.041 at 0x24beecb6208>\n"
     ]
    }
   ],
   "source": [
    "#using MOMA with COBRApy\n",
    "with model:\n",
    "    model.genes.b1723.knock_out()\n",
    "    model.genes.b3916.knock_out()\n",
    "    moma_result = cobra.flux_analysis.moma(model, fba_solution)\n",
    "    print('MOMA Result: ', moma_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "5f90bbc6",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ROOM Result:  <Solution 41.000 at 0x24beedc9988>\n"
     ]
    }
   ],
   "source": [
    "#using ROOM with COBRApy\n",
    "with model:\n",
    "    model.genes.b1723.knock_out()\n",
    "    model.genes.b3916.knock_out()\n",
    "    room_result = cobra.flux_analysis.room(model, fba_solution)\n",
    "    print('ROOM Result: ', room_result)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd66c161",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "#### Single Deletions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "613a478b",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Single gene and reaction delections can also be simulated with the flux analysis package of COBRApy. To do so the `cobra.flux_analysis.single_gene_deletion()` and `cobra.flux_analysis.single_reaction_deletion()` methods."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "1841f60e",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>growth</th>\n",
       "      <th>status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>{PIt2r}</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>{EX_pyr_e}</td>\n",
       "      <td>0.873922</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>{ME1}</td>\n",
       "      <td>0.873922</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>{FORt2}</td>\n",
       "      <td>0.873922</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>{PPS}</td>\n",
       "      <td>0.873922</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>{H2Ot}</td>\n",
       "      <td>0.383515</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>{ETOHt2r}</td>\n",
       "      <td>0.873922</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>{GND}</td>\n",
       "      <td>0.863813</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>{EX_etoh_e}</td>\n",
       "      <td>0.873922</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>{SUCDi}</td>\n",
       "      <td>0.814298</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>95 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            ids    growth   status\n",
       "0       {PIt2r}  0.000000  optimal\n",
       "1    {EX_pyr_e}  0.873922  optimal\n",
       "2         {ME1}  0.873922  optimal\n",
       "3       {FORt2}  0.873922  optimal\n",
       "4         {PPS}  0.873922  optimal\n",
       "..          ...       ...      ...\n",
       "90       {H2Ot}  0.383515  optimal\n",
       "91    {ETOHt2r}  0.873922  optimal\n",
       "92        {GND}  0.863813  optimal\n",
       "93  {EX_etoh_e}  0.873922  optimal\n",
       "94      {SUCDi}  0.814298  optimal\n",
       "\n",
       "[95 rows x 3 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#single reaction deletion\n",
    "reaction_deletion_results = cobra.flux_analysis.single_reaction_deletion(model)\n",
    "reaction_deletion_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "7ce5a73a",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>growth</th>\n",
       "      <th>status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>{b2097}</td>\n",
       "      <td>0.873922</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>{b0903}</td>\n",
       "      <td>0.873922</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>{b4090}</td>\n",
       "      <td>0.873922</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>{b2935}</td>\n",
       "      <td>0.873922</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>{b3603}</td>\n",
       "      <td>0.873922</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>{b0008}</td>\n",
       "      <td>0.873922</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>{b3213}</td>\n",
       "      <td>0.873922</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>{b1101}</td>\n",
       "      <td>0.873922</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>{b2297}</td>\n",
       "      <td>0.873922</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>{b0767}</td>\n",
       "      <td>0.863813</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>137 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         ids    growth   status\n",
       "0    {b2097}  0.873922  optimal\n",
       "1    {b0903}  0.873922  optimal\n",
       "2    {b4090}  0.873922  optimal\n",
       "3    {b2935}  0.873922  optimal\n",
       "4    {b3603}  0.873922  optimal\n",
       "..       ...       ...      ...\n",
       "132  {b0008}  0.873922  optimal\n",
       "133  {b3213}  0.873922  optimal\n",
       "134  {b1101}  0.873922  optimal\n",
       "135  {b2297}  0.873922  optimal\n",
       "136  {b0767}  0.863813  optimal\n",
       "\n",
       "[137 rows x 3 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#single gene deletion\n",
    "gene_deletion_results = cobra.flux_analysis.single_gene_deletion(model)\n",
    "gene_deletion_results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e800f7aa",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "It is worth noting that genes and reactions with a growth value equal to zero can be considered as essential genes or essential reactions, respectively."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f13be5b5",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "#### Double Delections"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61f39278",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Double gene and reaction delections can also be simulated with the flux analysis package of COBRApy. To do so the `cobra.flux_analysis.double_gene_deletion()` and `cobra.flux_analysis.double_reaction_deletion()` methods. These methods test the deletion of all the possible combinations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "4e16b218",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>growth</th>\n",
       "      <th>status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>{PFL, ICL}</td>\n",
       "      <td>0.873922</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>{ACt2r, NADH16}</td>\n",
       "      <td>0.206935</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>{EX_gln__L_e, D_LACt2}</td>\n",
       "      <td>0.873922</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>{ACt2r, RPE}</td>\n",
       "      <td>0.865716</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>{CO2t, PYK}</td>\n",
       "      <td>0.429326</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4555</th>\n",
       "      <td>{PGK, CS}</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4556</th>\n",
       "      <td>{EX_pyr_e, PPC}</td>\n",
       "      <td>0.870745</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4557</th>\n",
       "      <td>{FBA, PPS}</td>\n",
       "      <td>0.481972</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4558</th>\n",
       "      <td>{ADK1, MALS}</td>\n",
       "      <td>0.873922</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4559</th>\n",
       "      <td>{PFL, GLUt2r}</td>\n",
       "      <td>0.873922</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4560 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         ids    growth   status\n",
       "0                 {PFL, ICL}  0.873922  optimal\n",
       "1            {ACt2r, NADH16}  0.206935  optimal\n",
       "2     {EX_gln__L_e, D_LACt2}  0.873922  optimal\n",
       "3               {ACt2r, RPE}  0.865716  optimal\n",
       "4                {CO2t, PYK}  0.429326  optimal\n",
       "...                      ...       ...      ...\n",
       "4555               {PGK, CS}  0.000000  optimal\n",
       "4556         {EX_pyr_e, PPC}  0.870745  optimal\n",
       "4557              {FBA, PPS}  0.481972  optimal\n",
       "4558            {ADK1, MALS}  0.873922  optimal\n",
       "4559           {PFL, GLUt2r}  0.873922  optimal\n",
       "\n",
       "[4560 rows x 3 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#double reaction deletion\n",
    "double_reaction_deletion_results = cobra.flux_analysis.double_reaction_deletion(model)\n",
    "double_reaction_deletion_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "4dd61842",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>growth</th>\n",
       "      <th>status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>{b2914, b2282}</td>\n",
       "      <td>0.211663</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>{b3956, b3114}</td>\n",
       "      <td>0.870745</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>{b1852, b0734}</td>\n",
       "      <td>0.863813</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>{b3919, b0978}</td>\n",
       "      <td>0.704037</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>{b2417, b2458}</td>\n",
       "      <td>0.873922</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9448</th>\n",
       "      <td>{b2282, b1773}</td>\n",
       "      <td>0.211663</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9449</th>\n",
       "      <td>{b1524, b1380}</td>\n",
       "      <td>0.873922</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9450</th>\n",
       "      <td>{b1812, b2465}</td>\n",
       "      <td>0.873922</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9451</th>\n",
       "      <td>{b2282, b2925}</td>\n",
       "      <td>0.211663</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9452</th>\n",
       "      <td>{b2285, b2277}</td>\n",
       "      <td>0.211663</td>\n",
       "      <td>optimal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9453 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 ids    growth   status\n",
       "0     {b2914, b2282}  0.211663  optimal\n",
       "1     {b3956, b3114}  0.870745  optimal\n",
       "2     {b1852, b0734}  0.863813  optimal\n",
       "3     {b3919, b0978}  0.704037  optimal\n",
       "4     {b2417, b2458}  0.873922  optimal\n",
       "...              ...       ...      ...\n",
       "9448  {b2282, b1773}  0.211663  optimal\n",
       "9449  {b1524, b1380}  0.873922  optimal\n",
       "9450  {b1812, b2465}  0.873922  optimal\n",
       "9451  {b2282, b2925}  0.211663  optimal\n",
       "9452  {b2285, b2277}  0.211663  optimal\n",
       "\n",
       "[9453 rows x 3 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#double reaction deletion\n",
    "double_gene_deletion_results = cobra.flux_analysis.double_gene_deletion(model)\n",
    "double_gene_deletion_results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7db1b17",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "### Production envelopes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46da3428",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Production Envelopes can be used to show distinct phases of optimal growth using two different carbon substrates. If we want to access a phenotype phase plane to evaluate uptakes of Glucose and Oxygen:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "a765ac20",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>carbon_source</th>\n",
       "      <th>flux_minimum</th>\n",
       "      <th>carbon_yield_minimum</th>\n",
       "      <th>mass_yield_minimum</th>\n",
       "      <th>flux_maximum</th>\n",
       "      <th>carbon_yield_maximum</th>\n",
       "      <th>mass_yield_maximum</th>\n",
       "      <th>EX_glc__D_e</th>\n",
       "      <th>EX_o2_e</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EX_glc__D_e</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-10.0</td>\n",
       "      <td>-60.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EX_glc__D_e</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.072244</td>\n",
       "      <td>1.31005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-10.0</td>\n",
       "      <td>-56.842105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EX_glc__D_e</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.144488</td>\n",
       "      <td>2.62010</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-10.0</td>\n",
       "      <td>-53.684211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EX_glc__D_e</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.216732</td>\n",
       "      <td>3.93015</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-10.0</td>\n",
       "      <td>-50.526316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EX_glc__D_e</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.288975</td>\n",
       "      <td>5.24020</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-10.0</td>\n",
       "      <td>-47.368421</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  carbon_source  flux_minimum  carbon_yield_minimum  mass_yield_minimum  \\\n",
       "0   EX_glc__D_e           0.0                   0.0                 NaN   \n",
       "1   EX_glc__D_e           0.0                   0.0                 NaN   \n",
       "2   EX_glc__D_e           0.0                   0.0                 NaN   \n",
       "3   EX_glc__D_e           0.0                   0.0                 NaN   \n",
       "4   EX_glc__D_e           0.0                   0.0                 NaN   \n",
       "\n",
       "   flux_maximum  carbon_yield_maximum  mass_yield_maximum  EX_glc__D_e  \\\n",
       "0      0.000000               0.00000                 NaN        -10.0   \n",
       "1      0.072244               1.31005                 NaN        -10.0   \n",
       "2      0.144488               2.62010                 NaN        -10.0   \n",
       "3      0.216732               3.93015                 NaN        -10.0   \n",
       "4      0.288975               5.24020                 NaN        -10.0   \n",
       "\n",
       "     EX_o2_e  \n",
       "0 -60.000000  \n",
       "1 -56.842105  \n",
       "2 -53.684211  \n",
       "3 -50.526316  \n",
       "4 -47.368421  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#How to perform a production envolope\n",
    "prod_env = cobra.flux_analysis.production_envelope(model, [\"EX_glc__D_e\", \"EX_o2_e\"])\n",
    "prod_env.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f375970f",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Moreover, if we want to specify the carbon source, the `production_envelope()` can also return the carbon and mass yield. For instance, when the objective is to produce Acetate, it is possible to obtain a production envelope as follows and quickly plot the results:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "7880b2aa",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x24bf0b19cc8>"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAELCAYAAADJF31HAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8GearUAAAgAElEQVR4nO3dd3xV9f3H8dcni0CAACGsJKywdyCEIAgIgqBFioIMAZkBKo5S29r6a8WKtVpEa9Fq2KgsF6KioCJV0Exkz7DDDCvM7O/vj0QbICGXcG/OHZ/n45GHueece8/7i+HNN+eee44YY1BKKeX6vKwOoJRSyj600JVSyk1ooSullJvQQldKKTehha6UUm5CC10ppdyETYUuIn1FZLeIpIjI00Ws/72IbCr42iYiuSJSzf5xlVJKFUdKOg9dRLyBPUBvIBVIBIYZY3YUs31/4LfGmJ52zqqUUuombJmhRwEpxpj9xpgsYCkw4CbbDwOW2COcUkop2/nYsE0IcKTQ41SgU1EbikgFoC8wpZj1MUAMQEBAQIdmzZrdUlillPJ0ycnJp40xwUWts6XQpYhlxR2n6Q9sMMacLWqlMSYWiAWIjIw0SUlJNuxeKaXUz0TkUHHrbDnkkgqEFXocChwrZtuh6OEWpZSyhC2Fngg0FpEGIuJHfmmvvH4jEQkEugOf2DeiUkopW5R4yMUYkyMiU4DVgDcwzxizXUQmFax/q2DTgcAaY8xlh6VVSilVrBJPW3SUoo6hZ2dnk5qaSkZGhiWZlGfw9/cnNDQUX19fq6ModctEJNkYE1nUOlveFC0zqampVKpUifr16yNS1HuxSt0eYwxnzpwhNTWVBg0aWB1HKbtyqo/+Z2RkEBQUpGWuHEZECAoK0t8ClVtyqkIHtMyVw+nPmHJXTnXIRSnlWDm5eXyx7QTnr2RZF0KEiLAqtKxTWf9xtTMtdKU8xM7jF/jjh1vYkppudRQAwqqVp2/LWvRtVZuIsCp4eWm53y4tdDtasGABSUlJzJo1q0z3O378eKZOnUqLFi2K3aZHjx7MmDGDyMhr3xy3KvO9997L4sWLqVKlSpnu1xNl5uTyxrf7ePPbFALL+/L6sAjuCA+yME8eG/ae5ottx1nww0Fmf3+AWpX9uadlTfq2qk1Ug2p4a7mXiha6neTk5Fi27zlz5li279JatWqV1RE8wsbD5/jjB1vYe+oSAyNC+OuvWlA1wM/qWDzUMYyHOoaRfjWbtbtO8sXWEyxNPMLCHw8RFOBHn5a16NeqFp3Dg/D1drq3+pyW0xb6c59uZ8exC3Z9zRZ1KvNs/5Ylbrdo0SJmzJiBiNCmTRseeughpk+fTlZWFkFBQbz33nvUrFmTadOmcezYMQ4ePEj16tXp06cPR44coW/fvhw4cIDhw4fz7LPPAjBz5kzmzZsH5M+on3zySQ4ePEi/fv3o2rUrP/zwAyEhIXzyySeUL1/+hkz79u1j8ODBbNy4EYC9e/cydOhQkpOTr5l9r1mzhmeffZbMzEzCw8OZP38+FStWvOa15s+fz4svvkjt2rVp0qQJ5cqVK/bPYvTo0ZQvX55du3Zx6NAh5s+fz8KFC/nxxx/p1KkTCxYsAGDy5MkkJiZy9epVBg0axHPPPUd6ejpRUVGsXLmSpk2bMmzYMHr27MmECROoX78+SUlJXLp0ib59+9K1a1fi4uJo27YtY8aM4dlnn+XUqVO89957REVFMW3aNCpWrMhTTz0FQKtWrfjss88AbHq+p7mSlcMra/Ywb0P+7Hf+6I7c1ayG1bFuEFjel4ERoQyMCOVyZg7rdqfxxbbjrNx0lCUJhwks78vdzWvSr1Utujaujr+vt9WRnZrTFrpVtm/fzgsvvMCGDRuoXr06Z8+eRUSIi4tDRJgzZw4vv/wyr7zyCgDJycmsX7+e8uXLs2DBAhISEti2bRsVKlSgY8eO3HfffYgI8+fPJz4+HmMMnTp1onv37lStWpW9e/eyZMkSZs+ezUMPPcSHH37IiBEjbsgVHh5OYGAgmzZtol27dsyfP5/Ro0dfs83p06eZPn06X3/9NQEBAbz00kvMnDmTv/71r79sc/z4cZ599lmSk5MJDAzkrrvuIiIi4qZ/JufOnWPt2rWsXLmS/v37s2HDBubMmUPHjh1/yfPCCy9QrVo1cnNz6dWrF1u2bKFNmzbMmjWL0aNH88QTT3Du3DkmTJhww+unpKTw/vvvExsbS8eOHVm8eDHr169n5cqV/P3vf2fFihU3zXe7z3c3G1JO8/RHWzhy9iojo+vxh75NqeTv/B+iCijnw31tanNfm9pkZOfyfcFhmTU7TvDhxlQqlvPhrmY16NeqFve0rKWHZYrgtIVuy0zaEdauXcugQYOoXr06ANWqVWPr1q0MGTKE48ePk5WVdc0HUu6///5rZtS9e/cmKCj/+OQDDzzA+vXrEREGDhxIQEDAL8u///577r//fho0aEC7du0A6NChAwcPHiw22/jx45k/fz4zZ85k2bJlJCQkXLM+Li6OHTt20KVLFwCysrLo3LnzNdvEx8fTo0cPgoPzr745ZMgQ9uzZc9M/k/79+yMitG7dmpo1a9K6dWsAWrZsycGDB2nXrh3Lly8nNjaWnJwcjh8/zo4dO2jTpg29e/fm/fff59FHH2Xz5s1Fvn6DBg2uec1evXr9sr+b/XnY6/nuIv1qNi+u2snSxCM0qB7AsphoOjW07lj57fD39aZ3i5r0blGTrJw8fth3mi+3nWDNjpN8uvkYE7s35E/9mlsd0+k4baFbxRhzw6lUjz32GFOnTuX+++9n3bp1TJs27Zd1P5f0z65/rohws8srFD7c4e3tzdWrV4vd9sEHH+S5556jZ8+edOjQ4Zd/OApn7927N0uW3PyCl7d6qtjPGb28vK7J6+XlRU5ODgcOHGDGjBkkJiZStWpVRo8e/csHd/Ly8ti5cyfly5fn7NmzhIaGFvv61+/j59cH8PHxIS8v75ftCn8wyJbnu7s120/wfyu2ceZyFpO6h/Pk3Y3d5vCEn48XPZrWoEfTGkz/dR6/e38zCzYcZFyXBtSo7G91PKei7zZcp1evXixfvpwzZ84AcPbsWdLT0wkJCQFg4cKFN33+V199xdmzZ7l69SorVqygS5cudOvWjRUrVnDlyhUuX77Mxx9/zJ133nnL2fz9/bnnnnuYPHkyY8aMuWF9dHQ0GzZsICUlBYArV67cMPvu1KkT69at48yZM2RnZ/P+++/fco7rXbhwgYCAAAIDAzl58iRffPHFL+teffVVmjdvzpIlSxg7dizZ2dml2kf9+vV/ef9g48aNHDhw4LZzu4O0i5k8ungjMe8kE1SxHCt+04Wn+zVzmzK/no+3F1N7NyE3z/DGtylWx3E6OkO/TsuWLXnmmWfo3r073t7eREREMG3aNAYPHkxISAjR0dE3LZOuXbsycuRIUlJSGD58+C+nCY4ePfqXN+fGjx9PREREqQ4HPPzww3z00Uf06dPnhnXBwcEsWLCAYcOGkZmZCcD06dNp0qTJL9vUrl2badOm0blzZ2rXrk379u3Jzc295RyFtW3bloiICFq2bEnDhg1/OeSzZ88e5syZQ0JCApUqVaJbt25Mnz6d55577pb38eCDD7Jo0SLatWtHx44drxmTJzLGsGLTUZ77dAdXMnN5qk8TJnYP94gzQuoFBTA4MpQlCUeI6R5OSJUbTyLwVE51tcWdO3fSvLkeF7uZGTNmkJ6ezvPPP291FJfmyj9rR89f5ZmPt7Judxrt61bh5UFtaFSjktWxytTR81e565/reLBDKC8+0NrqOGXKZa62qG5u4MCB7Nu3j7Vr11odRVkk4cBZxi1IJNcYpvVvwcjO9T3ybI+QKuUZFhXGe/GHmdw9nLpBFayO5BS00J3Qo48+yoYNG65Z9sQTT/Dxxx87bJ8vvPDCDcfTBw8ezDPPPOOwfapb80PKacYtTKJOFX8WjIkirJpnl9ijdzViaeIR/vXNXl55qK3VcZyC0x1yadasmV6wRzmUMYZdu3a51CGX7/akMWFREvWDAnh3fCeCKxX/YTBP8sLnO5i7/gBfTe1OeHDFkp/gBm52yMWp3kHx9/fnzJkzNz3NT6nb8fMNLvz9Xed0t7W7TjJ+YRLhwRVZEhOtZV7IpO7h+Pt689rXe62O4hSc6pBLaGgoqamppKWlWR1FubGfb0HnCr7cdoLHlmykee3KLBobRZUK1l+HxZkEVSzHmC71eXPdPh69K5xmtSpbHclSTlXovr6+elswpQp8tuUYTyzdRJvQQBaOjaKyC3x83woT7mzIoh8O8epXe3h7ZJFHIjyGUx1yUUrl+/inVB5f8hMd6lblnXGdtMxvokoFP8bd2YDV20+y7ahzXOvdKjYVuoj0FZHdIpIiIk8Xs00PEdkkIttF5L/2jamU51iedISpyzcT3TCIBWM7UrGcU/0i7ZTGdm1AlQq+zPzq5tclcnclFrqIeANvAP2AFsAwEWlx3TZVgDeB+40xLYHBDsiqlNt7L/4Qf/hgC10bVWfe6I5U8NMyt0Vlf19iujVk7a5TJB86Z3Ucy9gyQ48CUowx+40xWcBSYMB12wwHPjLGHAYwxpyyb0yl3N+CDQd45uNt9GxWg9mjIt32eiyO8kjn+gQF+PGqB8/SbSn0EOBIocepBcsKawJUFZF1IpIsIqOKeiERiRGRJBFJ0jNZlPqf2O/2Me3THdzTsiZvjeigZV4KAeV8mNwjnPUpp4nbf8bqOJawpdCL+pTP9SeK+wAdgPuAe4C/iMgNV08yxsQaYyKNMZE/X49bKU83a+1e/r5qF79qU5tZw9vj56PnKpTWiOh61Kxcjplr9njk51ls+clJBcIKPQ4FjhWxzZfGmMvGmNPAd4B+FlepmzDGMPOrPcxYs4eBESG8NqSdR1wt0ZH8fb2ZclcjEg6eZX3KaavjlDlbfnoSgcYi0kBE/IChwMrrtvkEuFNEfESkAtAJ2GnfqEq5D2MML6/ezevf7GVwh1BmDG6Lj5a5XTzUMYyQKuWZ4YGz9BJ/gowxOcAUYDX5Jb3cGLNdRCaJyKSCbXYCXwJbgARgjjFmm+NiK+W6jDFM/3wn/1m3j4c71eWlB9t45BUTHaWcjzeP9WzE5iPnWbvLs87PcKqLcynl7vLyDNM+3c6iHw8x+o76PNu/hV6MzgGyc/O4e+Z/qVjOh0+ndMXLjf7BdJmLcynlzvLyDH/+eCuLfjxETLeGWuYO5OvtxRO9GrP92AVWbz9hdZwyo4WuVBnIzTP8/oMtLE08wpS7GvGnfnqZaEcb0C6E8OAAXv16D7l5nnEsXQtdKQfLyc1j6vJNfLgxlam9m/DUPU21zMuAt5fw295N2HPyEp9tuf7EPPekha6UA2Xn5vH40p/4ZNMx/ti3GY/3amx1JI9yb6vaNKtVide+3ktObp7VcRxOC10pB8nMyWXyuxtZtfUE/3dfcyb3CLc6ksfx8hKm9m7CgdOX+fino1bHcTgtdKUcICM7l4nvJPP1zpP8bUBLxt/Z0OpIHqt3i5q0CQ3kX9/sJSvHvWfpWuhK2dnVrFzGL0ziv3vSePGB1ozqXN/qSB5NJH+WnnruKu8nHyn5CS5MC10pO7qcmcOYBQn8sO80/xzUlmFRda2OpIDuTYLpUK8qs9amkJGda3Uch9FCV8pOLmZk88i8BBIPnuPVIe0Y1ME17lvqCUSE3/VuwvH0DJYkHLY6jsNooStlB+lXshkxN4FNR87z72ERDGh3/RWmldXuaFSdzg2DeOPbfVzNcs9Zuha6Urfp3OUshs+JY8exdN58uD33tq5tdSRVjN/1acLpS5ks+vGg1VEcQgtdqdtw+lImw2bHsffUJWJHRdKnZS2rI6mbiKxfje5Ngnnrv/u4lJljdRy700JXqpROXchgWGwcB89cZt4jHbmraQ2rIykbTO3dhHNXslmw4YDVUexOC12pUjiRnsHQ2DiOnr/KgjFRdG1c3epIykZtw6pwd/OaxH63n/Sr2VbHsSstdKVuUeq5Kzz09o+cupjJorFRRDcMsjqSukVTezfhQkYOc7/fb3UUu9JCV+oWHD5zhSFvx3HuShbvjIsisn41qyOpUmhRpzL3ta7N3PUHOHs5y+o4dqOFrpSNDpy+zJDYH7mclcOSCdFE1K1qdSR1G568uzFXsnN5+7t9VkexGy10pWyQcuoiD739I1k5eSyZEE2rkECrI6nb1LhmJX7dLoSFPxzk1MUMq+PYhRa6UiXYdeICQ96OA2BpTDTNa1e2OJGylyd6NSY71/Cfde4xS9dCV+omth1NZ1hsHL7eXiyLiaZxzUpWR1J2VL96AA+2D+G9+MMcT79qdZzbpoWuVDE2HznP8NlxVPDzYdnEaBoGV7Q6knKAx3o2xhjDrLUpVke5bTYVuoj0FZHdIpIiIk8Xsb6HiKSLyKaCr7/aP6pSZSf50FlGzIknsIIvyyZGUy8owOpIykHCqlVgSMcwlicd4cjZK1bHuS0lFrqIeANvAP2AFsAwEWlRxKbfG2PaFXz9zc45lSoz8fvPMHJuAtUrlWP5xM6EVq1gdSTlYFPuaoyI8O+1e62OcltsmaFHASnGmP3GmCxgKTDAsbGUssaGlNM8Mj+BOlXKsywmmtqB5a2OpMpArUB/RnSqx4cbj3Lg9GWr45SaLYUeAhS+zUdqwbLrdRaRzSLyhYi0LOqFRCRGRJJEJCktLa0UcZVynHW7TzF2QSL1gwJYGhNNjcr+VkdSZWhyj3D8vL3419d7rI5SarYUuhSxzFz3eCNQzxjTFvg3sKKoFzLGxBpjIo0xkcHBwbeWVCkH+nrHSWIWJRMeXJHFE6KpXrGc1ZFUGQuuVI5H7qjPJ5uPsefkRavjlIothZ4KhBV6HAocK7yBMeaCMeZSwferAF8R0asVKZfw5bbjTHo3mea1K7FkQjTVAvysjqQsMrFbQwL8fHjNRWfpthR6ItBYRBqIiB8wFFhZeAMRqSUiUvB9VMHrnrF3WKXsbeXmYzy6+CfahlXhnfGdCKzga3UkZaGqAX6M7VKfVVtPsP1YutVxblmJhW6MyQGmAKuBncByY8x2EZkkIpMKNhsEbBORzcDrwFBjzPWHZZRyKh8mp/Lk0p/oUK8qC8dGUdlfy1zBuDsbUtnfh1e/cr1Zuo8tGxUcRll13bK3Cn0/C5hl32hKOc6yxMM8/dFWOjcMYs4jkVTws+mvgvIAgeV9ienWkBlr9rDpyHnahVWxOpLN9JOiyuO8E3eIP364lW6Ng5k3uqOWubrB6C4NqFrBl1fW7LY6yi3RQlceZd76A/xlxTbubl6D2FEd8Pf1tjqSckIVy/kwuUc43+89TeLBs1bHsZkWuvIYb/13H3/7bAd9W9bizYc7UM5Hy1wVb2R0fYIrlWPG6t24yluCWujKI7z+zV7+8cUu+retw6zhEfj56I++urnyft482iOc+ANn+WGfa5y0pz/Vyq0ZY3hlzW5mfrWHByJCeG1IO3y89cde2WZoVF1qB/rzyhrXmKXrT7ZyW8YY/vHlLv69NoUhkWH8c3BbvL2K+uCzUkXz9/VmSs9GbDx8nnW7nf9yJVroyi0ZY/jbZzt4+7/7GRFdlxcfaK1lrkplcIcwwqqVZ+ZXe5x+lq6FrtxOXp7hL59sY/6Gg4zpUp/nB7TCS8tclZKfjxeP92zM1qPprNlx0uo4N6WFrtxKXp7hzx9v5d24w0zs3pC//qoFBVelUKrUBkaE0LB6ADPX7CEvz3ln6Vroym3k5hme+mAzSxOP8HjPRjzdt5mWubILH28vnri7MbtPXuTzrcetjlMsLXTlFrJz83hy2SY+2niUqb2bMLVPUy1zZVf929ShSc2KvPr1HnJy86yOUyQtdOXysnLyeHzJT3y6+RhP92vG470aWx1JuSEvL+G3dzdhf9plPtl0rOQnWEALXbm0zJxcfvNeMl9sO8FfftWCSd3DrY6k3Ng9LWvRsk5l/vXNXrKdcJauha5cVkZ2LjGLkvl65yme/3UrxnVtYHUk5ea8vISpvZtw+OwVPkhOtTrODbTQlUu6kpXDuIWJfLc3jZcebM3I6HpWR1IeomezGrQLq8K/v9lLZk6u1XGuoYWuXM6lzBxGz0/kx31nmDGoLUM61rU6kvIgIsLv+jThWHoGSxOOWB3nGlroyqVcyMjmkXkJJB86x2tDI3iwQ6jVkZQH6tqoOlH1q/HGtylkZDvPLF0LXbmM9CvZjJwTz+Yj55k1LIL729axOpLyUD/P0k9dzOTduENWx/mFFrpyCWcvZzF8Thw7j1/krREd6Ne6ttWRlIfr1DCIro2q8+a6fVzOzLE6DqCFrlzA6UuZDJ8dx95Tl4gd1YG7W9S0OpJSAEzt04Szl7NY8MNBq6MANha6iPQVkd0ikiIiT99ku44ikisig+wXUXmyUxcyGBobx8Ezl5k/uiM9mtawOpJSv2hftyo9m9Ug9rv9XMjItjpOyYUuIt7AG0A/oAUwTERaFLPdS8Bqe4dUnul4+lWGxMZx7PxVFoyJokuj6lZHUuoGU3s3If1qNnO/P2B1FJtm6FFAijFmvzEmC1gKDChiu8eAD4FTdsynPFTquSsMeTuOtIuZvDMuiuiGQVZHUqpIrUIC6duyFvPWH+Dc5SxLs9hS6CFA4ZMtUwuW/UJEQoCBwFs3eyERiRGRJBFJSktz/rt/KGscOnOZIW/Hcf5KFu+O70SHetWsjqTUTf22dxMuZeUQ+/1+S3PYUuhFXbLu+gsCvwb80Rhz0xMyjTGxxphIY0xkcHCwrRmVB9mfdokhb8dxOSuHxROiaRdWxepISpWoaa1K9G9ThwUbDnL6UqZlOWwp9FQgrNDjUOD6S41FAktF5CAwCHhTRH5tl4TKY+w9eZEhsXFk5+axNCaaViGBVkdSymZP3N2YzJxc/rNun2UZbCn0RKCxiDQQET9gKLCy8AbGmAbGmPrGmPrAB8BvjDEr7J5Wua2dxy8wNDYOgKUx0TSrVdniRErdmvDgijzQPpR34w5x8kKGJRlKLHRjTA4whfyzV3YCy40x20VkkohMcnRA5f62HU1n2Ow4fL29WBYTTeOalayOpFSpPNGrMbl5hje+TbFk/z62bGSMWQWsum5ZkW+AGmNG334s5Sk2HTnPqLnxVPL3ZfGETtQLCrA6klKlFlatAoMjw1iScJiYbg0JrVqhTPevnxRVlkk6eJYRc+KpUsGPZROjtcyVW3isZyMEYdbasp+la6ErS8TtP8OoeQnUqFSOZROjy3wmo5Sj1KlSnuGd6vJ+cioHT18u031roasyt37vaUbPTyCkSnmWxkRTO7C81ZGUsqvf9AjH11t4/Zu9ZbpfLXRVpr7dfYqxCxOpHxTAkphoalT2tzqSUnZXo7I/ozrXZ8Wmo6Sculhm+9VCV2Xmqx0nmbgomcY1KrJ4QjTVK5azOpJSDjOxW0P8fb159euym6Vroasy8cXW40x+N5nmdSqzeHw01QL8rI6klEMFVSzH2C4N+HzLcXYev1Am+9RCVw73yaajTFnyE23DqvDuuCgCK/haHUmpMjHhzoZU8vdh5ld7ymR/WujKoT5ITuW3yzYRWa8qi8ZGUclfy1x5jsAKvky4syFf7TjJltTzDt+fFrpymKUJh/n9B5u5I7w6C8ZEEVDOps+xKeVWxnSpT5UKvmUyS9dCVw7xzo8HefqjrXRrHMycRyIp7+dtdSSlLFHJ35eJ3cJZtzuN5ENnHbovLXRld3PXH+Avn2zn7uY1iR3VAX9fLXPl2R65ox7VK/rxyhrHztK10JVd/WfdPp7/bAf9WtXizYfbU85Hy1ypCn4+TO7RiB/2neGHfacdth8tdGU3r3+zl5e+3MX9bevw72ER+Pnoj5dSP3u4U11qVfZn5po9GHP9PYLsQ//GqdtmjGHG6t3M/GoPD7QP4dUh7fDx1h8tpQrz9/Xm0Z6NSDp0ju/2OmaWrn/r1G0xxvCPL3Yx69sUhnYMY8agtnh7FXXXQqXUkMgwwqqVZ9Nhx5zCqOeRqVIzxvC3z3Ywf8NBRkbX47n7W+KlZa5Usfx8vFj9ZDcq+DmmerXQVank5Rn+8sk23os/zLiuDfi/+5ojomWuVEkcVeagha5KITfP8KePtrA8KZVJ3cP5Y9+mWuZKOQEtdHVLcnLz+MMHW/jop6M83qsxv727sZa5Uk5CC13ZLDs3j98u28RnW47zVJ8mTOnZ2OpISqlCtNCVTbJy8nhsyUZWbz/Jn+9tRky3cKsjKaWuY9NpiyLSV0R2i0iKiDxdxPoBIrJFRDaJSJKIdLV/VGWVjOxcJr+bzOrtJ3m2fwstc6WcVIkzdBHxBt4AegOpQKKIrDTG7Ci02TfASmOMEZE2wHKgmSMCq7KVkZ1LzDvJfLcnjem/bsWI6HpWR1JKFcOWGXoUkGKM2W+MyQKWAgMKb2CMuWT+91nWAMAxn2tVZepKVg5jFyTy/d40Xn6wjZa5Uk7OlkIPAY4UepxasOwaIjJQRHYBnwNji3ohEYkpOCSTlJaWVpq8qoxcysxh9LxE4vafYeZDbXmoY5jVkZRSJbCl0Is6J+2GGbgx5mNjTDPg18DzRb2QMSbWGBNpjIkMDg6+taSqzFzIyGbU3HiSD5/jX0MjGBgRanUkpZQNbCn0VKDw9CwUOFbcxsaY74BwEal+m9mUBdKvZDNiTjxbj6bzxvAI+retY3UkpZSNbCn0RKCxiDQQET9gKLCy8AYi0kgKPl0iIu0BP+CMvcMqxzp7OYths+PYdfwib43oQN9Wta2OpJS6BSWe5WKMyRGRKcBqwBuYZ4zZLiKTCta/BTwIjBKRbOAqMMQ46oK/yiHSLmYyYk48B89cZvYjkXRvoofElHI1YlXvRkZGmqSkJEv2ra518kIGw2fHcex8BnMfieSORnq0TClnJSLJxpjIotbpJ0U93LHzVxk+O460i5ksGNORTg2DrI6klColLXQPduTsFYbPieP85WwWjetEh3pVrY6klLoNWuge6tCZywyfHc/FjGzeHd+JtmFVrI6klLpNWugeaF/aJYbPjiMrJ4/FE6JpFRJodSSllB1ooXuYvScvMmx2PGBYEhNNs1qVrY6klLITLXQPsl2Mqg0AABESSURBVPP4BUbMicfbS1g8IZpGNSpZHUkpZUda6B5i29F0RsyNp7yvN4snRNOgeoDVkZRSdqaF7gF+OnyOUfMSqOzvy9KYaMKqVbA6klLKAWy6wYVyXUkHzzJybgJVK/ixbKKWuVLuTGfobuzHfWcYtzCRWpX9WTwhmlqB/lZHUko5kM7Q3dT6vacZsyCBkCrlWTpRy1wpT6AzdDf07a5TTHw3mYbVA3hvfCeCKpazOpJSqgxoobuZNdtP8OjijTStVYl3xnaiaoCf1ZGUUmVEC92NrNp6nMeX/ETLkEAWjY0isLyv1ZGUUmVIj6G7iU82HeWxJT/RLqwK747TMlfKE2mhu4EPklN5ctkmOtavysKxUVTy1zJXyhPpIRcXtyThMH/+eCtdwqsze1Qk5f28rY6klLKIztBd2KIfD/Knj7bSvUkwcx7RMlfK0+kM3UXN+X4/0z/fSe8WNZk1PIJyPlrmSnk6LXQX9Oa6FF7+cjf3tq7Fv4ZG4Outv2gppWw85CIifUVkt4ikiMjTRax/WES2FHz9ICJt7R9VGWP419d7efnL3QxoV4fXtcyVUoWUOEMXEW/gDaA3kAokishKY8yOQpsdALobY86JSD8gFujkiMCeyhjDjDW7eePbfTzYPpSXB7XB20usjqWUciK2TO+igBRjzH5jTBawFBhQeANjzA/GmHMFD+OAUPvG9GzGGP6+aidvfLuPYVFh/FPLXClVBFsKPQQ4UuhxasGy4owDvridUOp/jDE89+kOZn9/gFGd6/HCr1vjpWWulCqCLW+KFtUepsgNRe4iv9C7FrM+BogBqFu3ro0RPVdenuH/PtnG4vjDjO/agGfua46IlrlSqmi2zNBTgbBCj0OBY9dvJCJtgDnAAGPMmaJeyBgTa4yJNMZEBgcHlyavx8jNM/zxwy0sjj/Mb3qEa5krpUpkS6EnAo1FpIGI+AFDgZWFNxCRusBHwEhjzB77x/QsObl5/G75Jt5PTuWJXo35/T1NtcyVUiUq8ZCLMSZHRKYAqwFvYJ4xZruITCpY/xbwVyAIeLOgeHKMMZGOi+2+snPzeHLZJj7fcpzf39OUR+9qZHUkpZSLEGOKPBzucJGRkSYpKcmSfTurrJw8HluykdXbT/LMvc2Z0K2h1ZGUUk5GRJKLmzDrJ0WdREZ2Lr95byNrd51iWv8WjO7SwOpISikXo4XuBK5m5RLzThLf7z3N3we2ZngnPQNIKXXrtNAtdiUrh3ELkog7cIaXB7Xhociwkp+klFJF0EK30KXMHMbMTyD50DlmPtSWgRH6AVulVOlpoVsk/Wo2o+cnsCU1ndeHRfCrNnWsjqSUcnFa6BY4fyWLkXMT2HXiAm8+3J57WtayOpJSyg1ooZexM5cyGTE3gX1pl3h7ZAd6NqtpdSSllJvQQi9Dpy5mMGJOPIfOXGHOqEi6NdHLHyil7EcLvYycvJDBsNlxHD+fwfzRHbmjUXWrIyml3IwWehk4dv4qw2fHkXYxk4Vjo4hqUM3qSEopN6SF7mBHzl5h2Ow40q9m8874TrSvW9XqSEopN6WF7kAHT19m+Ow4Lmflsnh8NK1DA62OpJRyY1roDpJy6hLDZ8eRk2dYMiGaFnUqWx1JKeXmtNAdYPeJizw8Jw4QlkyIpmmtSlZHUkp5AC10O9tx7AIj5sbj4yUsnhBNoxoVrY6klPIQttyxSNloS+p5hs2Ow9/Hi+UTO2uZK6XKlM7Q7WTj4XM8MjeBwAq+LJkQTVi1ClZHUkp5GC10O0g4cJYx8xOoXqkciydEE1KlvNWRlFIeSAv9Nv2w7zTjFiRRu4o/SyZEU7Oyv9WRlFIeSo+h34bv9qQxZn4iYdXKsyyms5a5UspSOkMvpbW7TjLpnY2E16jIu+OiCKpYzupISikPZ9MMXUT6ishuEUkRkaeLWN9MRH4UkUwRecr+MZ3Ll9tOMPGdZJrWqsSSCZ20zJVSTqHEGbqIeANvAL2BVCBRRFYaY3YU2uws8Djwa4ekdCKfbTnGE0s30SY0kAVjoggs72t1JKWUAmyboUcBKcaY/caYLGApMKDwBsaYU8aYRCDbARmdxsc/pfL4kp9oX7cKi8ZqmSulnIsthR4CHCn0OLVg2S0TkRgRSRKRpLS0tNK8hGWWJx1h6vLNdGoQxMKxUVTy1zJXSjkXWwpdilhmSrMzY0ysMSbSGBMZHOw6d+t5L/4Qf/hgC10bVWfe6I5U8NP3kpVSzseWZkoFwgo9DgWOOSaO81mw4QDTPt1Bz2Y1ePPh9vj7elsdSSmlimTLDD0RaCwiDUTEDxgKrHRsLOcw+7v9TPt0B31a1OStER20zJVSTq3EGboxJkdEpgCrAW9gnjFmu4hMKlj/lojUApKAykCeiDwJtDDGXHBgdod649sU/rl6N/e1qc1rQ9rh662fwVJKOTebDgYbY1YBq65b9lah70+QfyjG5RljeO3rvfzrm70MjAjhn4Pa4KNlrpRyAfruXiHGGF5evZv/rNvH4A6h/OPBNnh7FfWesFJKOR8t9ALGGKZ/vpO56w8wvFNdpg9ohZeWuVLKhWihA3l5hmmfbmfRj4cYfUd9nu3fAhEtc6WUa/H4Qs/LMzyzYitLEo4Q060hf+rXTMtcKeWSPLrQc/MMf/hgCx9uTGXKXY34XZ8mWuZKKZflsYWek5vH797fzCebjjG1dxMe79XY6khKKXVbPLLQs3PzeGLpT6zaeoI/9G3Kb3o0sjqSUkrdNo8r9MycXKYs/omvdpzk/+5rzvg7G1odSSml7MKjCj0jO5fJ7ybz7e40/jagJaM617c6klJK2Y3HFPrVrFwmLEpiw77TvPhAa4ZF1bU6klJK2ZVHFPrlzBzGLUwk/sBZ/jmoLYM6uMVVCpRS6hpuX+gXM7IZMz+Rn46c57Uh7RjQrlT35lBKKafn1oWefjWbR+YlsO1oOv8eFsG9rWtbHUkppRzGbQv93OUsRs6LZ/eJi7z5cHv6tKxldSSllHIotyz005cyGTEnnv2nLxM7MpK7mtWwOpJSSjmc2xX6qYsZPDw7niPnrjD3kUjubOw69y5VSqnb4VaFfiI9g+Gz4zhxIYP5o6PoHB5kdSSllCozblPoR89fZfjsOM5cymLR2Cgi61ezOpJSSpUptyj0w2euMGx2HBcysnlnXBQRdataHUkppcqcyxf6gdOXGT47jqvZuSweH03r0ECrIymllCVcutBTTl1k+Ox4cvIMi8dH06JOZasjKaWUZWy6nb2I9BWR3SKSIiJPF7FeROT1gvVbRKS9/aNea/eJiwyNjSPPwNIYLXOllCqx0EXEG3gD6Ae0AIaJSIvrNusHNC74igH+Y+ec19h2NJ2hsT/i7SUsmxhNk5qVHLk7pZRyCbbM0KOAFGPMfmNMFrAUGHDdNgOARSZfHFBFRBzyOfvNR84zfHYcFfx8WD6xM+HBFR2xG6WUcjm2FHoIcKTQ49SCZbe6DSISIyJJIpKUlpZ2q1kBMEBo1QosjYmmXlBAqV5DKaXckS2FXtRdk00ptsEYE2uMiTTGRAYHl+4TnO3CqvDZY10Jq1ahVM9XSil3ZUuhpwJhhR6HAsdKsY3deHkV9e+HUkp5NlsKPRFoLCINRMQPGAqsvG6blcCogrNdooF0Y8xxO2dVSil1EyWeh26MyRGRKcBqwBuYZ4zZLiKTCta/BawC7gVSgCvAGMdFVkopVRSbPlhkjFlFfmkXXvZWoe8N8Kh9oymllLoVNn2wSCmllPPTQldKKTehha6UUm5CC10ppdyE5L+facGORdKAQ6V8enXgtB3jWEnH4pzcZSzuMg7QsfysnjGmyE9mWlbot0NEkowxkVbnsAcdi3Nyl7G4yzhAx2ILPeSilFJuQgtdKaXchKsWeqzVAexIx+Kc3GUs7jIO0LGUyCWPoSullLqRq87QlVJKXUcLXSml3ITLFbqIPFZww+rtIvJyoeV/KrhJ9W4RucfKjLYQkWkiclRENhV83VtonUuNBUBEnhIRIyLVCy1zqXGIyPMFNznfJCJrRKROoXWuNpZ/isiugvF8LCJVCq1ztbEMLvj7nicikdetc7Wx9C3ImiIiT9t9B8YYl/kC7gK+BsoVPK5R8N8WwGagHNAA2Ad4W523hLFMA54qYrkrjiWM/MsrHwKqu/A4Khf6/nHgLRceSx/Ap+D7l4CXXHgszYGmwDogstBylxoL+Zcf3wc0BPwKsrew5z5cbYY+GfiHMSYTwBhzqmD5AGCpMSbTGHOA/OuyR1mU8Xa54lheBf7AtbcddLlxGGMuFHoYwP/G44pjWWOMySl4GEf+XcTANcey0xizu4hVrjaWKCDFGLPfGJMFLCV/DHbjaoXeBLhTROJF5L8i0rFguU03qXZCUwp+JZ4nIlULlrnUWETkfuCoMWbzdatcahw/E5EXROQI8DDw14LFLjmWQsYCXxR87+pjKczVxuLwvDbd4KIsicjXQK0iVj1Dft6qQDTQEVguIg2x8SbVZa2EsfwHeJ78nM8Dr5D/F8/pxlLCOP5M/q/3NzytiGVO/f/EGPOJMeYZ4BkR+RMwBXgWFx1LwTbPADnAez8/rYjtXWIsRT2tiGWWj+UmHJ7X6QrdGHN3cetEZDLwkck/IJUgInnkX+SmTG9SbaubjaUwEZkNfFbw0OnGUtw4RKQ1+ccuN4sI5GfdKCJROOE4wPb/J8Bi4HPyC90lxyIijwC/AnoV/J0BFx1LMZxyLDfh+LxWv1Fwi28qTAL+VvB9E/J/fRGgJde+ObIfJ35zpCB/7ULf/5b8Y4G44lgKjeMg/3tT1OXGATQu9P1jwAcuPJa+wA4g+LrlLjeWQtnXce2boi41FvIn0PsLsv78pmhLe+7D6WboJZgHzBORbUAW8IjJ/5PaLiLLyf8BzgEeNcbkWpjTFi+LSDvyf+U6CEwEMPk34Ha1sdzARcfxDxFpCuSRf8bOzzdCd8WxzCK/6L4q+O0pzhgzyRXHIiIDgX8DwcDnIrLJGHOPq43FGJMjIlPIPyPMG5hnjNluz33oR/+VUspNuNpZLkoppYqhha6UUm5CC10ppdyEFrpSSrkJLXSllHITWuhKKeUmtNCVWxCR3EKXIt4kIk+LiLeIJItIt0LbrRGRwbf42hVE5POCy9FuF5F/2H8ESt0+PQ9duQURuWSMqVjE8k7AHKA9MAgYbYy5petmi0gFoJMx5lsR8QO+Af5ujPmihKcqVaZ0hq7cmjEmHviB/OvP/x149Gbbi8gwEdkqIttE5KWC17hijPm24PssYCP/uxxtUa8RLCIfikhiwVcXOw1HqZvSGbpyCyKSC2wttOhFY8yygnXVyL/uz2sm/2qKxb1GHfKvHd4BOAesAV43xqwotE0V8gv9bmPM/mJeZzHwpjFmvYjUBVYbY5rf1gCVsoGrXctFqeJcNca0K2ZdNyAdaFXCa3QE1hlj0gBE5L2C564oeOwDLCG/5Iss8wJ3Ay0KrqECUFlEKhljLto0EqVKSQ+5KLcmIgHAy0BPILjwvVuL2ryEl4sF9hpjXithOy+gszGmXcFXiJa5Kgta6Mrd/RVYbozZBfwGeFVE/IvZNh7oLiLVRcQbGAb8F0BEpgOBwJM27HMN+TfHoOC5xf3moJRd6TF05RaKOIb+JbCI/MMlbY0xVwu2ex04Y4x5rpjXGQ78ifzZ+ipjzB9EJJT8Y/C7gMyCTWcZY+YU8xrVgTfIv7mxD/CdMWbSbQ5RqRJpoSullJvQQy5KKeUm9CwX5ZFEJJ78O/oUNtIYs7Wo7Yt5jWeA6z91+r4x5oXbzadUaeghF6WUchN6yEUppdyEFrpSSrkJLXSllHITWuhKKeUm/h+oDPrWaJpRYQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#obtain a plot of a production envelope using glucose as carbon source and optimizing Acetate production\n",
    "prod_env = cobra.flux_analysis.production_envelope(model, [\"EX_o2_e\"], objective=\"EX_ac_e\", carbon_sources=\"EX_glc__D_e\")\n",
    "prod_env.plot(kind='line', x='EX_o2_e', y='carbon_yield_maximum')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2807d7bf",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "Finally, the production envelope can also be used obtain a flux variation plot. This can be achieved as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "878664b8",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'Biomass_Ecoli_core'",
     "output_type": "error",
     "traceback": [
      "\u001B[1;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[1;31mKeyError\u001B[0m                                  Traceback (most recent call last)",
      "\u001B[1;32m<ipython-input-34-f8bf16800972>\u001B[0m in \u001B[0;36m<module>\u001B[1;34m\u001B[0m\n\u001B[0;32m      1\u001B[0m \u001B[1;31m#obtain a flux varariation plot\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[1;32m----> 2\u001B[1;33m \u001B[0mprod_env_co2\u001B[0m \u001B[1;33m=\u001B[0m \u001B[0mcobra\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0mflux_analysis\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0mproduction_envelope\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0mmodel\u001B[0m\u001B[1;33m,\u001B[0m \u001B[1;33m[\u001B[0m\u001B[1;34m\"Biomass_Ecoli_core\"\u001B[0m\u001B[1;33m]\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0mobjective\u001B[0m\u001B[1;33m=\u001B[0m\u001B[1;33m[\u001B[0m\u001B[1;34m'EX_co2_e'\u001B[0m\u001B[1;33m]\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0m\u001B[0;32m      3\u001B[0m \u001B[0mprod_env_co2\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0mplot\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0mkind\u001B[0m\u001B[1;33m=\u001B[0m\u001B[1;34m'line'\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0mx\u001B[0m\u001B[1;33m=\u001B[0m\u001B[1;34m'Biomass_Ecoli_core'\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0my\u001B[0m\u001B[1;33m=\u001B[0m\u001B[1;33m[\u001B[0m\u001B[1;34m'flux_minimum'\u001B[0m\u001B[1;33m,\u001B[0m \u001B[1;34m'flux_maximum'\u001B[0m\u001B[1;33m]\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n",
      "\u001B[1;32mc:\\users\\bisbii\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\cobra\\flux_analysis\\phenotype_phase_plane.py\u001B[0m in \u001B[0;36mproduction_envelope\u001B[1;34m(model, reactions, objective, carbon_sources, points, threshold)\u001B[0m\n\u001B[0;32m    102\u001B[0m \u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0;32m    103\u001B[0m     \"\"\"\n\u001B[1;32m--> 104\u001B[1;33m     \u001B[0mreactions\u001B[0m \u001B[1;33m=\u001B[0m \u001B[0mmodel\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0mreactions\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0mget_by_any\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0mreactions\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0m\u001B[0;32m    105\u001B[0m     \u001B[0mobjective\u001B[0m \u001B[1;33m=\u001B[0m \u001B[0mmodel\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0msolver\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0mobjective\u001B[0m \u001B[1;32mif\u001B[0m \u001B[0mobjective\u001B[0m \u001B[1;32mis\u001B[0m \u001B[1;32mNone\u001B[0m \u001B[1;32melse\u001B[0m \u001B[0mobjective\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0;32m    106\u001B[0m     \u001B[0mdata\u001B[0m \u001B[1;33m=\u001B[0m \u001B[0mdict\u001B[0m\u001B[1;33m(\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n",
      "\u001B[1;32mc:\\users\\bisbii\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\cobra\\core\\dictlist.py\u001B[0m in \u001B[0;36mget_by_any\u001B[1;34m(self, iterable)\u001B[0m\n\u001B[0;32m    106\u001B[0m         \u001B[1;32mif\u001B[0m \u001B[1;32mnot\u001B[0m \u001B[0misinstance\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0miterable\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0mlist\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m:\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0;32m    107\u001B[0m             \u001B[0miterable\u001B[0m \u001B[1;33m=\u001B[0m \u001B[1;33m[\u001B[0m\u001B[0miterable\u001B[0m\u001B[1;33m]\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[1;32m--> 108\u001B[1;33m         \u001B[1;32mreturn\u001B[0m \u001B[1;33m[\u001B[0m\u001B[0mget_item\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0mitem\u001B[0m\u001B[1;33m)\u001B[0m \u001B[1;32mfor\u001B[0m \u001B[0mitem\u001B[0m \u001B[1;32min\u001B[0m \u001B[0miterable\u001B[0m\u001B[1;33m]\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0m\u001B[0;32m    109\u001B[0m \u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0;32m    110\u001B[0m     def query(\n",
      "\u001B[1;32mc:\\users\\bisbii\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\cobra\\core\\dictlist.py\u001B[0m in \u001B[0;36m<listcomp>\u001B[1;34m(.0)\u001B[0m\n\u001B[0;32m    106\u001B[0m         \u001B[1;32mif\u001B[0m \u001B[1;32mnot\u001B[0m \u001B[0misinstance\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0miterable\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0mlist\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m:\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0;32m    107\u001B[0m             \u001B[0miterable\u001B[0m \u001B[1;33m=\u001B[0m \u001B[1;33m[\u001B[0m\u001B[0miterable\u001B[0m\u001B[1;33m]\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[1;32m--> 108\u001B[1;33m         \u001B[1;32mreturn\u001B[0m \u001B[1;33m[\u001B[0m\u001B[0mget_item\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0mitem\u001B[0m\u001B[1;33m)\u001B[0m \u001B[1;32mfor\u001B[0m \u001B[0mitem\u001B[0m \u001B[1;32min\u001B[0m \u001B[0miterable\u001B[0m\u001B[1;33m]\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0m\u001B[0;32m    109\u001B[0m \u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0;32m    110\u001B[0m     def query(\n",
      "\u001B[1;32mc:\\users\\bisbii\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\cobra\\core\\dictlist.py\u001B[0m in \u001B[0;36mget_item\u001B[1;34m(item)\u001B[0m\n\u001B[0;32m     98\u001B[0m                 \u001B[1;32mreturn\u001B[0m \u001B[0mself\u001B[0m\u001B[1;33m[\u001B[0m\u001B[0mitem\u001B[0m\u001B[1;33m]\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0;32m     99\u001B[0m             \u001B[1;32melif\u001B[0m \u001B[0misinstance\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0mitem\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0mstr\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m:\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[1;32m--> 100\u001B[1;33m                 \u001B[1;32mreturn\u001B[0m \u001B[0mself\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0mget_by_id\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0mitem\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0m\u001B[0;32m    101\u001B[0m             \u001B[1;32melif\u001B[0m \u001B[0mitem\u001B[0m \u001B[1;32min\u001B[0m \u001B[0mself\u001B[0m\u001B[1;33m:\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0;32m    102\u001B[0m                 \u001B[1;32mreturn\u001B[0m \u001B[0mitem\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n",
      "\u001B[1;32mc:\\users\\bisbii\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\cobra\\core\\dictlist.py\u001B[0m in \u001B[0;36mget_by_id\u001B[1;34m(self, id)\u001B[0m\n\u001B[0;32m     72\u001B[0m     \u001B[1;32mdef\u001B[0m \u001B[0mget_by_id\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0mself\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0mid\u001B[0m\u001B[1;33m:\u001B[0m \u001B[0mUnion\u001B[0m\u001B[1;33m[\u001B[0m\u001B[0mObject\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0mstr\u001B[0m\u001B[1;33m]\u001B[0m\u001B[1;33m)\u001B[0m \u001B[1;33m->\u001B[0m \u001B[0mObject\u001B[0m\u001B[1;33m:\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0;32m     73\u001B[0m         \u001B[1;34m\"\"\"Return the element with a matching id.\"\"\"\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[1;32m---> 74\u001B[1;33m         \u001B[1;32mreturn\u001B[0m \u001B[0mlist\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0m__getitem__\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0mself\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0mself\u001B[0m\u001B[1;33m.\u001B[0m\u001B[0m_dict\u001B[0m\u001B[1;33m[\u001B[0m\u001B[0mid\u001B[0m\u001B[1;33m]\u001B[0m\u001B[1;33m)\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0m\u001B[0;32m     75\u001B[0m \u001B[1;33m\u001B[0m\u001B[0m\n\u001B[0;32m     76\u001B[0m     \u001B[1;32mdef\u001B[0m \u001B[0mlist_attr\u001B[0m\u001B[1;33m(\u001B[0m\u001B[0mself\u001B[0m\u001B[1;33m,\u001B[0m \u001B[0mattribute\u001B[0m\u001B[1;33m:\u001B[0m \u001B[0mstr\u001B[0m\u001B[1;33m)\u001B[0m \u001B[1;33m->\u001B[0m \u001B[0mlist\u001B[0m\u001B[1;33m:\u001B[0m\u001B[1;33m\u001B[0m\u001B[1;33m\u001B[0m\u001B[0m\n",
      "\u001B[1;31mKeyError\u001B[0m: 'Biomass_Ecoli_core'"
     ]
    }
   ],
   "source": [
    "#obtain a flux varariation plot\n",
    "prod_env_co2 = cobra.flux_analysis.production_envelope(model, [\"Biomass_Ecoli_core\"], objective=['EX_co2_e'])\n",
    "prod_env_co2.plot(kind='line', x='Biomass_Ecoli_core', y=['flux_minimum', 'flux_maximum'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e50d254",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "### Pathway Visualization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4a21891",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "COBRApy does not include any pathway visualization method. Nevertheless, independent tools that work with cobra models can be use for that goal. One of such tools is named Escher.\n",
    "\n",
    "Escher allows the visualization of metabolic pathway maps. Escher maps can be build using the `escher.Build()` function, which requires the model and a FBA solution.\n",
    "\n",
    "To obtain the metabolic map for the model in question, Escher can be run as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d0526c7",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "import escher\n",
    "\n",
    "builder = escher.Builder(map_name='e_coli_core.Core metabolism', model=model, reaction_data=fba_solution.fluxes)\n",
    "builder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cbb2360",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}